Flake: OVN cluster-manager EgressIP Operations syncEgressIP for IPv4 [It] should not update valid assignment

### Which jobs are flaking?

```
cd go-controller/pkg/clustermanager; for i in {1..100}; do echo === $i ===; time go test -count 1 .; if [ $? != 0 ]; then break; fi; done
```

Yields:
```
------------------------------
• [FAILED] [0.021 seconds]
OVN cluster-manager EgressIP Operations syncEgressIP for IPv4 [It] should not update valid assignment
/home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager/egressip_controller_test.go:4003

  [FAILED] Expected
      <string>: node2
  to equal
      <string>: node1
  In [It] at: /home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager/egressip_controller_test.go:4082 @ 09/30/25 14:43:36.448
(...)
Summarizing 1 Failure:
  [FAIL] OVN cluster-manager EgressIP Operations syncEgressIP for IPv4 [It] should not update valid assignment
  /home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager/egressip_controller_test.go:4082

Ran 92 of 92 Specs in 14.974 seconds
FAIL! -- 91 Passed | 1 Failed | 0 Pending | 0 Skipped
```

### Which tests are flaking?

```
[root@ovnkubernetes clustermanager]# for i in {1..100}; do echo === $i ===; time go test -count 1 -ginkgo.focus '.*OVN cluster-manager EgressIP Operations syncEgressIP for IPv4 should not update valid assignment.*' .; if [ $? != 0 ]; then break; fi; done
=== 1 ===
Running Suite: Cluster Manager Operations Suite - /home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager
==========================================================================================================================
Random Seed: 1759236455

Will run 1 of 92 specs
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSI0930 14:47:35.189424   17211 factory.go:531] Starting watch factory
I0930 14:47:35.189596   17211 envvar.go:172] "Feature gate default state" feature="ClientsPreferCBOR" enabled=false
I0930 14:47:35.189630   17211 envvar.go:172] "Feature gate default state" feature="InformerResourceVersion" enabled=false
I0930 14:47:35.189649   17211 envvar.go:172] "Feature gate default state" feature="InOrderInformers" enabled=true
I0930 14:47:35.189659   17211 envvar.go:172] "Feature gate default state" feature="WatchListClient" enabled=false
I0930 14:47:35.189682   17211 envvar.go:172] "Feature gate default state" feature="ClientsAllowCBOR" enabled=false
I0930 14:47:35.189741   17211 reflector.go:357] "Starting reflector" type="*v1.Service" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.189757   17211 reflector.go:403] "Listing and watching" type="*v1.Service" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.189906   17211 reflector.go:430] "Caches populated" type="*v1.Service" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.189989   17211 reflector.go:357] "Starting reflector" type="*v1.EndpointSlice" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.190018   17211 reflector.go:403] "Listing and watching" type="*v1.EndpointSlice" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.190106   17211 reflector.go:430] "Caches populated" type="*v1.EndpointSlice" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.191579   17211 reflector.go:357] "Starting reflector" type="*v1.Node" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.191695   17211 reflector.go:403] "Listing and watching" type="*v1.Node" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.191898   17211 reflector.go:430] "Caches populated" type="*v1.Node" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.195180   17211 factory.go:1890] *v1.Node informer cache synced successfully
I0930 14:47:35.195210   17211 factory.go:1890] *v1.Service informer cache synced successfully
I0930 14:47:35.195215   17211 factory.go:1890] *v1.EndpointSlice informer cache synced successfully
I0930 14:47:35.195283   17211 reflector.go:357] "Starting reflector" type="*v1.EgressIP" resyncPeriod="0s" reflector="pkg/crd/egressip/v1/apis/informers/externalversions/factory.go:140"
I0930 14:47:35.195429   17211 reflector.go:403] "Listing and watching" type="*v1.EgressIP" reflector="pkg/crd/egressip/v1/apis/informers/externalversions/factory.go:140"
I0930 14:47:35.195494   17211 reflector.go:430] "Caches populated" type="*v1.EgressIP" reflector="pkg/crd/egressip/v1/apis/informers/externalversions/factory.go:140"
I0930 14:47:35.200345   17211 factory.go:1890] *v1.EgressIP informer cache synced successfully
I0930 14:47:35.200392   17211 factory.go:636] Watch Factory start up complete, took: 10.972951ms
I0930 14:47:35.200461   17211 egressip_controller.go:1747] Patching status on EgressIP egressip: [{add /metadata/annotations map[k8s.ovn.org/egressip-mark:50000]}]
I0930 14:47:35.200972   17211 obj_retry.go:509] Add event received for *v1.EgressIP egressip
E0930 14:47:35.201042   17211 egressip_controller.go:1437] Allocator error: EgressIP: egressip has mistmach with status vs cache for node: node1 with IP: 192.168.126.101
I0930 14:47:35.201122   17211 egressip_controller.go:1187] Current assignments are: map[192.168.126.111:{Node:node1 Name:bogus2} 192.168.126.68:{Node:node2 Name:bogus3}]
I0930 14:47:35.201146   17211 egressip_controller.go:1189] Will attempt assignment for egress IP: 192.168.126.101
I0930 14:47:35.201175   17211 egressip_controller.go:1297] Attempting assignment on egress node: &{egressIPConfig:0xc000625d40 mgmtIPs:[] allocations:map[192.168.126.68:bogus3] healthClient:0xc00083a040 isReady:true isReachable:true isEgressAssignable:true name:node2}
I0930 14:47:35.201218   17211 egressip_controller.go:1340] Successful assignment of egress IP: 192.168.126.101 to network 192.168.126.0/24 on node: &{egressIPConfig:0xc000625d40 mgmtIPs:[] allocations:map[192.168.126.101:egressip 192.168.126.68:bogus3] healthClient:0xc00083a040 isReady:true isReachable:true isEgressAssignable:true name:node2}
I0930 14:47:35.201237   17211 egressip_controller.go:877] Setting egress IP node allocation - node: node2, EIP name: egressip, IP: 192.168.126.101
I0930 14:47:35.201255   17211 egressip_controller.go:1747] Patching status on EgressIP egressip: [{add /metadata/annotations map[k8s.ovn.org/egressip-mark:50000]} {replace /status {[{node2 192.168.126.101}]}}]
I0930 14:47:35.201365   17211 obj_retry.go:548] Creating *v1.EgressIP egressip took: 364.904µs
I0930 14:47:35.201391   17211 factory.go:1326] Added *v1.EgressIP event handler 1
I0930 14:47:35.201606   17211 obj_retry.go:565] Update event received for resource *v1.EgressIP, old object is equal to new: false
I0930 14:47:35.201641   17211 obj_retry.go:617] Update event received for *v1.EgressIP egressip
I0930 14:47:35.201697   17211 egressip_controller.go:877] Setting egress IP node allocation - node: node2, EIP name: egressip, IP: 192.168.126.101
I0930 14:47:35.201712   17211 obj_retry.go:565] Update event received for resource *v1.EgressIP, old object is equal to new: false
I0930 14:47:35.201723   17211 obj_retry.go:617] Update event received for *v1.EgressIP egressip
I0930 14:47:35.201739   17211 egressip_controller.go:877] Setting egress IP node allocation - node: node2, EIP name: egressip, IP: 192.168.126.101
I0930 14:47:35.202110   17211 watch.go:218] "Stopping fake watcher"
I0930 14:47:35.202162   17211 reflector.go:363] "Stopping reflector" type="*v1.EgressIP" resyncPeriod="0s" reflector="pkg/crd/egressip/v1/apis/informers/externalversions/factory.go:140"
I0930 14:47:35.202221   17211 handler.go:192] Sending *v1.EgressIP event handler 1 for removal
I0930 14:47:35.202267   17211 handler.go:210] Removed *v1.EgressIP event handler 1
I0930 14:47:35.203115   17211 watch.go:218] "Stopping fake watcher"
I0930 14:47:35.203156   17211 reflector.go:363] "Stopping reflector" type="*v1.Node" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.203171   17211 watch.go:218] "Stopping fake watcher"
I0930 14:47:35.203194   17211 reflector.go:363] "Stopping reflector" type="*v1.Service" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.203200   17211 watch.go:218] "Stopping fake watcher"
I0930 14:47:35.203249   17211 reflector.go:363] "Stopping reflector" type="*v1.EndpointSlice" resyncPeriod="0s" reflector="k8s.io/client-go/informers/factory.go:160"
I0930 14:47:35.203399   17211 factory.go:642] Stopping watch factory
I0930 14:47:35.203431   17211 obj_retry.go:436] Stop channel got triggered: will stop retrying failed objects of type *v1.EgressIP

------------------------------
• [FAILED] [0.037 seconds]
OVN cluster-manager EgressIP Operations syncEgressIP for IPv4 [It] should not update valid assignment
/home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager/egressip_controller_test.go:4003

  [FAILED] Expected
      <string>: node2
  to equal
      <string>: node1
  In [It] at: /home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager/egressip_controller_test.go:4082 @ 09/30/25 14:47:35.201
------------------------------
SSSS

Summarizing 1 Failure:
  [FAIL] OVN cluster-manager EgressIP Operations syncEgressIP for IPv4 [It] should not update valid assignment
  /home/akaris/development/ovn-kubernetes/go-controller/pkg/clustermanager/egressip_controller_test.go:4082

Ran 1 of 92 Specs in 0.039 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 91 Skipped
--- FAIL: TestClusterManager (0.04s)
FAIL
exit status 1
FAIL    github.com/ovn-org/ovn-kubernetes/go-controller/pkg/clustermanager      0.058s

real    0m1.690s
user    0m2.613s
sys     0m0.865s
```

### Since when has it been flaking?

no idea, but it's pretty persistent and easy to reproduce

### Reason for failure (if possible)

_No response_

### Anything else we need to know?

_No response_

--- a/go-controller/pkg/clustermanager/egressip_controller_test.go
+++ b/go-controller/pkg/clustermanager/egressip_controller_test.go
@@ -3,8 +3,12 @@ package clustermanager
 import (
 	"context"
 	"fmt"
+	"maps"
 	"net"
+	"reflect"
+	"slices"
 	"strconv"
+	"strings"
 	"sync"
 	"time"
 
@@ -294,6 +298,62 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 			return ""
 		}
 		return nodes[0]
+	}
+
+	egressIPsMatch := func(expectedEgressIPs []egressipv1.EgressIP) func() bool {
+		return func() (result bool) {
+			egressIPList, err := fakeClusterManagerOVN.fakeClient.EgressIPClient.K8sV1().EgressIPs().List(context.TODO(), metav1.ListOptions{})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			gotEgressIPs := egressIPList.Items
+
+			defer func() {
+				if !result {
+					ginkgo.GinkgoWriter.Printf("Mismatch in egressIP lists, got: %+v, expected: %+v\n", gotEgressIPs, expectedEgressIPs)
+				}
+			}()
+
+			if len(expectedEgressIPs) != len(gotEgressIPs) {
+				return false
+			}
+
+			numEquals := 0
+			for _, expectedEgressIP := range expectedEgressIPs {
+				for _, gotEgressIP := range gotEgressIPs {
+					if expectedEgressIP.Name != gotEgressIP.Name {
+						continue
+					}
+
+					specExpectedEIP := slices.Clone(expectedEgressIP.Spec.EgressIPs)
+					specGotEgressIP := slices.Clone(gotEgressIP.Spec.EgressIPs)
+					slices.Sort(specExpectedEIP)
+					slices.Sort(specGotEgressIP)
+					if !reflect.DeepEqual(specExpectedEIP, specGotEgressIP) {
+						return false
+					}
+
+					statusExpectedEgressIP := slices.Clone(expectedEgressIP.Status.Items)
+					statusGotEgressIP := slices.Clone(gotEgressIP.Status.Items)
+					sortFunc := func(a, b egressipv1.EgressIPStatusItem) int {
+						return strings.Compare(a.EgressIP, b.EgressIP)
+					}
+					slices.SortFunc(statusExpectedEgressIP, sortFunc)
+					slices.SortFunc(statusGotEgressIP, sortFunc)
+					if !reflect.DeepEqual(statusExpectedEgressIP, statusGotEgressIP) {
+						return false
+					}
+					numEquals++
+				}
+			}
+			return len(expectedEgressIPs) == numEquals
+		}
+	}
+
+	readAllocations := func(nodeName string) map[string]string {
+		fakeClusterManagerOVN.eIPC.nodeAllocator.Lock()
+		defer fakeClusterManagerOVN.eIPC.nodeAllocator.Unlock()
+		node, ok := fakeClusterManagerOVN.eIPC.nodeAllocator.cache[nodeName]
+		gomega.Expect(ok).To(gomega.BeTrue(), fmt.Sprintf("node %s missing from allocator cache", nodeName))
+		return maps.Clone(node.allocations)
 	}
 
 	ginkgo.BeforeEach(func() {
@@ -3264,15 +3324,21 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 
 	ginkgo.Context("syncEgressIP for dual-stack", func() {
 
-		ginkgo.It("should not update valid assignments", func() {
-			// FIXME(mk): this test doesn't ensure that the EIP status does not get patched during the test run
-			// and therefore is an invalid test to test that the status is not patched
+		// This test validates that if the allocator cache contains valid entries that match
+		// the egress IP status items, no reassignment shall happen.
+		// In order to do so, it does 2 comparisons:
+		// a) take egressIP that's passed to the cluster manager and compare it to the list of EgressIPs after running
+		//    WatchEgressIP.
+		// b) take egressNode.allocations before running WatchEgressIP and make sure that they haven't changed.
+		ginkgo.It("should not reallocate if cache matches EgressIP status", func() {
 			app.Action = func(*cli.Context) error {
 				config.IPv6Mode = true
 				config.IPv4Mode = true
 				egressIPv4 := "192.168.126.101"
 				egressIPv6 := "0:0:0:0:0:feff:c0a8:8e0d"
-				node1IPv6 := "0:0:0:0:0:feff:c0a8:8e0c/64"
+				node1IPv6 := "0:0:0:0:0:feff:c0a8:8e0b/64"
+				node2IPv6 := "0:0:0:0:0:feff:c0a8:8e0c/64"
+				node1IPv4 := "192.168.126.50/16"
 				node2IPv4 := "192.168.126.51/16"
 
 				node1 := corev1.Node{
@@ -3318,8 +3384,15 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 					},
 				}
 
-				egressNode1 := setupNode(node1Name, []string{node1IPv6}, map[string]string{})
-				egressNode2 := setupNode(node2Name, []string{node2IPv4}, map[string]string{"192.168.126.102": "bogus3"})
+				egressNode1 := setupNode(node1Name, []string{node1IPv4, node1IPv6}, map[string]string{
+					net.ParseIP(egressIPv6).String(): egressIPName, // Cache matches eIP Status.
+				})
+				egressNode2 := setupNode(node2Name, []string{node2IPv4, node2IPv6}, map[string]string{
+					"192.168.126.102": "bogus3",
+					egressIPv4:        egressIPName, // Cache matches eIP Status.
+				})
+				egressNode1OriginalAllocations := maps.Clone(egressNode1.allocations)
+				egressNode2OriginalAllocations := maps.Clone(egressNode2.allocations)
 
 				eIP := egressipv1.EgressIP{
 					ObjectMeta: newEgressIPMeta(egressIPName),
@@ -3350,11 +3423,11 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 
 				_, err := fakeClusterManagerOVN.eIPC.WatchEgressIP()
 				gomega.Expect(err).NotTo(gomega.HaveOccurred())
-
-				gomega.Eventually(getEgressIPStatusLen(egressIPName)).Should(gomega.Equal(2))
-				egressIPs, nodes := getEgressIPStatus(egressIPName)
-				gomega.Expect(nodes).To(gomega.ConsistOf(eIP.Status.Items[0].Node, eIP.Status.Items[1].Node))
-				gomega.Expect(egressIPs).To(gomega.ConsistOf(eIP.Status.Items[0].EgressIP, eIP.Status.Items[1].EgressIP))
+				gomega.Eventually(egressIPsMatch([]egressipv1.EgressIP{eIP})).Should(gomega.BeTrue())
+				gomega.Eventually(readAllocations).WithArguments(egressNode1.name).Should(
+					gomega.Equal(egressNode1OriginalAllocations))
+				gomega.Eventually(readAllocations).WithArguments(egressNode2.name).Should(
+					gomega.Equal(egressNode2OriginalAllocations))
 				return nil
 			}
 
@@ -4000,7 +4073,13 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 			gomega.Expect(err).NotTo(gomega.HaveOccurred())
 		})
 
-		ginkgo.It("should not update valid assignment", func() {
+		// This test validates that if the allocator cache contains valid entries that match
+		// the egress IP status items, no reassignment shall happen.
+		// In order to do so, it does 2 comparisons:
+		// a) take egressIP that's passed to the cluster manager and compare it to the list of EgressIPs after running
+		//    WatchEgressIP.
+		// b) take egressNode.allocations before running WatchEgressIP and make sure that they haven't changed.
+		ginkgo.It("should not reallocate if cache matches EgressIP status", func() {
 			app.Action = func(*cli.Context) error {
 				egressIP1 := "192.168.126.101"
 				node1IPv4 := "192.168.126.12/24"
@@ -4049,8 +4128,15 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 					},
 				}
 
-				egressNode1 := setupNode(node1Name, []string{"192.168.126.12/24"}, map[string]string{"192.168.126.111": "bogus2"})
-				egressNode2 := setupNode(node2Name, []string{"192.168.126.51/24"}, map[string]string{"192.168.126.68": "bogus3"})
+				egressNode1 := setupNode(node1Name, []string{"192.168.126.12/24"}, map[string]string{
+					"192.168.126.111": "bogus2",
+					egressIP1:         egressIPName, // Cache matches eIP Status.
+				})
+				egressNode2 := setupNode(node2Name, []string{"192.168.126.51/24"}, map[string]string{
+					"192.168.126.68": "bogus3",
+				})
+				egressNode1OriginalAllocations := maps.Clone(egressNode1.allocations)
+				egressNode2OriginalAllocations := maps.Clone(egressNode2.allocations)
 
 				eIP := egressipv1.EgressIP{
 					ObjectMeta: newEgressIPMeta(egressIPName),
@@ -4077,10 +4163,12 @@ var _ = ginkgo.Describe("OVN cluster-manager EgressIP Operations", func() {
 
 				_, err := fakeClusterManagerOVN.eIPC.WatchEgressIP()
 				gomega.Expect(err).NotTo(gomega.HaveOccurred())
-				gomega.Eventually(getEgressIPStatusLen(egressIPName)).Should(gomega.Equal(1))
-				egressIPs, nodes := getEgressIPStatus(egressIPName)
-				gomega.Expect(nodes[0]).To(gomega.Equal(egressNode1.name))
-				gomega.Expect(egressIPs[0]).To(gomega.Equal(egressIP1))
+
+				gomega.Eventually(egressIPsMatch([]egressipv1.EgressIP{eIP})).Should(gomega.BeTrue())
+				gomega.Eventually(readAllocations).WithArguments(egressNode1.name).Should(
+					gomega.Equal(egressNode1OriginalAllocations))
+				gomega.Eventually(readAllocations).WithArguments(egressNode2.name).Should(
+					gomega.Equal(egressNode2OriginalAllocations))
 				return nil
 			}
 

